sympy/solvers/ode/single.py

```diff
@@ -141,6 +163,53 @@ def is_autonomous(self):
         syms = self.eq.subs(self.func, u).free_symbols
         return x not in syms
 
+    def get_linear_coefficients(self, eq, func, order):
+        r"""
+        Matches a differential equation to the linear form:
+
+        .. math:: a_n(x) y^{(n)} + \cdots + a_1(x)y' + a_0(x) y + B(x) = 0
+
+        Returns a dict of order:coeff terms, where order is the order of the
```

Comment:
Can this function just use the `_lineq2dict` function defined here:
https://github.com/sympy/sympy/blob/6a57794ca820ac055106e7f421f12e99e5aac832/sympy/polys/matrices/linsolve.py#L171-L172

Reply:
```python
>>> _lin_eq2dict(eq, {f(x),f(x).diff(x),f(x).diff(x,2),f(x).diff(x,3)})
(x - sin(x), {f(x): -1, Derivative(f(x), x): cos(x) + 2, Derivative(f(x), (x, 2)): x, Derivative(f(x), (x, 3)): 1})
```
This also looks fine, only output format is changed. so should I change to use this function?

Reply:
Yes. that should simplify it a bit.

The docstring should be rewritten just to say that it gets the coefficients of the ODE. Is it more convenient for the calling code to have a dict with e.g. `f(x)` as the key or just a list of coefficients in order of derivatives?

Reply:
I think dictionary is good as it provides more readability. If you want me to change to list, I will do that

Reply:
@oscarbenjamin `_lin_eq2dict()` does not check whether the given ODE is linear. so should I keep `nth_linear_match` code only?
```python
>>> eq=f(x).diff(x, 3) + 2*f(x).diff(x) +x*f(x).diff(x, 2) + cos(x)*f(x).diff(x) + x - f(x) -sin(f(x))
>>> _nth_linear_match(eq, f(x), 3)
None
>>> _lin_eq2dict(eq,{f(x),f(x).diff(x),f(x).diff(x,2),f(x).diff(x,3)})
(x - sin(f(x)), {f(x): -1, Derivative(f(x), x): cos(x) + 2, Derivative(f(x), (x, 2)): x, Derivative(f(x), (x, 3)): 1})
```

Reply:
Oh yes, I forgot about that.

You could check if the rhs or any of the values of the dict has `f(x)` like:
```python
In [1]: sin(f(x)).has(f(x))
Out[1]: True
```